torch/csrc/jit/script/compiler.cpp

```diff
@@ -445,7 +488,29 @@ at::optional<std::vector<Value*>> tryMatchSchema(
         v.value = createStack(graph, loc, unpacked);
       }
 
-      if(!v.value->type()->isSubtypeOf(*arg.type)) {
+      bool skip_subtype_check = false;
+
+      // implicit conversion from Python scalar to Tensor if applicable.
+      // Occurs when we math tensors & scalars (ie, `x + 3.1` or `3.1 * 2.7`)
+      if (arg.type->isSubtypeOf(*DynamicType::get()) && isPythonScalar(v.value->type())) {
+        // Does not looked at named arguments or default args.
+        // This is only used in math functions (+-*/) so this is okay.
+        auto repr_val = determinePythonMathType(schema, inputs);
+        if (repr_val) {
+          // cast to tensor's type (ScalarType and backend)
+          std::vector<Value*> type_as_args({ v.value, repr_val.value().value });
+          auto* castNode = graph.insertNode(
+            graph.create(aten::type_as, type_as_args)
+            ->setSourceLocation(std::make_shared<SourceRange>(loc)));
+          v.value = castNode->output();
+        } else {
+          // No casting needs to be done, however, FloatType / IntType are NOT
+          // tensor types so we should skip the below subtype check.
+          skip_subtype_check = true;
```

Comment:
This skip logic is going to be fragile. Ideally, I'd introduce and insert a prim::cast node here whose only purpose is to mark the change in type, but it would require more fundamental changes downstream. For now, let's keep this as is, but we should reconsider if omitting explicit casts is every a source of bugs.